## Project description
In our project we'll prepare a report for a bank’s loan division. we'll need to find out if a customer’s marital status and number of children have an impact on whether they will default on a loan.

## Open the data file and have a look at the general information. 

[Start with importing the libraries and loading the data. You may realise that you need additional libraries as you go, which is totally fine - just make sure to update this section when you do.]

In [1]:
# Loading all the libraries
import pandas as pd
import numpy as np

# Load the data
try:
    df = pd.read_csv(r"C:\Users\tomsh\Desktop\Python Projects\Project 1 - Data preprocessing\credit_scoring_eng.csv")
except:
    df = pd.read_csv('/datasets/credit_scoring_eng.csv')
    


## Task 1. Data exploration

**Description of the data**
- `children` - the number of children in the family
- `days_employed` - work experience in days
- `dob_years` - client's age in years
- `education` - client's education
- `education_id` - education identifier
- `family_status` - marital status
- `family_status_id` - marital status identifier
- `gender` - gender of the client
- `income_type` - type of employment
- `debt` - was there any debt on loan repayment
- `total_income` - monthly income
- `purpose` - the purpose of obtaining a loan

[Now let's explore our data. You'll want to see how many columns and rows it has, look at a few rows to check for potential issues with the data.]

In [2]:
# Let's see how many rows and columns our dataset has

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [3]:
# let's print the first N rows

df.head()

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

In [4]:
df.describe().T

count          mean            std           min  \
children          21525.0      0.538908       1.381587     -1.000000   
days_employed     19351.0  63046.497661  140827.311974 -18388.949901   
dob_years         21525.0     43.293380      12.574584      0.000000   
education_id      21525.0      0.817236       0.548138      0.000000   
family_status_id  21525.0      0.972544       1.420324      0.000000   
debt              21525.0      0.080883       0.272661      0.000000   
total_income      19351.0  26787.568355   16475.450632   3306.762000   

                           25%           50%           75%            max  
children              0.000000      0.000000      1.000000      20.000000  
days_employed     -2747.423625  -1203.369529   -291.095954  401755.400475  
dob_years            33.000000     42.000000     53.000000      75.000000  
education_id          1.000000      1.000000      1.000000       4.000000  
family_status_id      0.000000      0.000000      1.000000       4.000000  
debt                  0.000000      0.000000      0.000000       1.000000  
total_income      16488.504500  23202.870000  32549.611000  362496.645000

[Describe what you see and notice in your printed data sample. Are there any issues that may need further investigation and changes?] 

1) There is 2 columns with missing values (could be symmetric) - need further investigation.
2) Some nummerical columns are include negative values - need further investigation.

In [5]:
# Let's look at the filtered table with missing values in the the first column with missing data
df[df.days_employed.isnull()]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

In [6]:
# Let's apply multiple conditions for filtering data and look at the number of rows in the filtered table.
df[(df.days_employed.isnull())&(df.total_income.isnull())]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

**Intermediate conclusion**


The Number of rows in the filtered table match the number of missing values (2174). Our conclusion is that the missing values (in all the columns with missing values) are symmetric.

**My next steps:**

1) calculate percentage of the missing values compare to whole dataset.

2) check for a dependencies of the missing values with identified charactristics - i.e: maybe missing values for total income are caused by employment such as retiree?

In [7]:
# percentage of the missing values compared to the whole dataset
print('{:.0%}'.format(len(df[df.days_employed.isnull()])/len(df))) 

10%


In [8]:
# checking distribution
df[df.total_income.isnull()]['income_type'].value_counts()

employee         1105
business          508
retiree           413
civil servant     147
entrepreneur        1
Name: income_type, dtype: int64

In [9]:
df[df.total_income.isnull()]['income_type'].value_counts()/len(df)

employee         0.051336
business         0.023600
retiree          0.019187
civil servant    0.006829
entrepreneur     0.000046
Name: income_type, dtype: float64

In [10]:
# Checking the distribution in the whole dataset
df['income_type'].value_counts()

employee                       11119
business                        5085
retiree                         3856
civil servant                   1459
entrepreneur                       2
unemployed                         2
paternity / maternity leave        1
student                            1
Name: income_type, dtype: int64

In [11]:
df['income_type'].value_counts()/len(df)

employee                       0.516562
business                       0.236237
retiree                        0.179141
civil servant                  0.067782
entrepreneur                   0.000093
unemployed                     0.000093
paternity / maternity leave    0.000046
student                        0.000046
Name: income_type, dtype: float64

 **conclusion**

There is no pattern that points up a strong relation for my theory wich was that retiree clients are caused the missing values for total income. 

# 3  Data transformation
[Let's go through each column to see what issues we may have in them.]

[Begin with removing duplicates and fixing educational information if required.]

In [12]:
# Let's see all values in education column to check if and what spellings will need to be fixed
df.education.unique()

array(["bachelor's degree", 'secondary education', 'Secondary Education',
       'SECONDARY EDUCATION', "BACHELOR'S DEGREE", 'some college',
       'primary education', "Bachelor's Degree", 'SOME COLLEGE',
       'Some College', 'PRIMARY EDUCATION', 'Primary Education',
       'Graduate Degree', 'GRADUATE DEGREE', 'graduate degree'],
      dtype=object)

In [13]:
# Fix the registers if required
df.education = df.education.str.lower()

In [14]:
# Checking all the values in the column to make sure we fixed them
df.education.unique()

array(["bachelor's degree", 'secondary education', 'some college',
       'primary education', 'graduate degree'], dtype=object)

In [15]:
# Let's see the distribution of values in the `children` column
df.children.value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [16]:
df.children.value_counts()/len(df)

 0     0.657329
 1     0.223833
 2     0.095470
 3     0.015331
 20    0.003531
-1     0.002184
 4     0.001905
 5     0.000418
Name: children, dtype: float64

In [17]:
# [fix the data based on your decision]
replace_dict = {-1:1,20:2}
df['children'] = df['children'].replace(replace_dict)

In [18]:
# Checking the `children` column again to make sure it's all fixed
df.children.value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

In [19]:
# Find problematic data in `days_employed`, if they exist, and calculate the percentage
df.days_employed.value_counts()

-327.685916     1
-1580.622577    1
-4122.460569    1
-2828.237691    1
-2636.090517    1
               ..
-7120.517564    1
-2146.884040    1
-881.454684     1
-794.666350     1
-3382.113891    1
Name: days_employed, Length: 19351, dtype: int64

**The problematic data may caused to some technical issues.
We will use the abs() method to convert negative values to positive values.**

In [20]:
df.days_employed.value_counts()/len(df)

-327.685916     0.000046
-1580.622577    0.000046
-4122.460569    0.000046
-2828.237691    0.000046
-2636.090517    0.000046
                  ...   
-7120.517564    0.000046
-2146.884040    0.000046
-881.454684     0.000046
-794.666350     0.000046
-3382.113891    0.000046
Name: days_employed, Length: 19351, dtype: float64

In [21]:
# Address the problematic values, if they exist
df['days_employed'] = df['days_employed'].abs()

In [22]:
# Check the result - make sure it's fixed
df['days_employed'].min()

24.14163324048118

In [23]:
# Check the `dob_years` for suspicious values and count the percentage
df['dob_years'].describe()

count    21525.000000
mean        43.293380
std         12.574584
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

In [24]:
df[df['dob_years']<=18]['dob_years'].value_counts()/len(df)

0    0.004692
Name: dob_years, dtype: float64

**There is a very small percentage of problematic (dob_years == 0) rows in the dob_years column.
In this case we can remove those rows from our data.**

In [25]:
# Address the issues in the `dob_years` column, if they exist
df = df[df.dob_years != 0]

In [26]:
# Check the result - make sure it's fixed
df[df['dob_years']==0]['dob_years'].count()

0

In [27]:
# Let's see the values for the family status column

df.family_status.value_counts()

married              12331
civil partnership     4156
unmarried             2797
divorced              1185
widow / widower        955
Name: family_status, dtype: int64

In [28]:
# Let's see the values in the gender column
df.gender.value_counts()

F      14164
M       7259
XNA        1
Name: gender, dtype: int64

In [29]:
df[df['gender']=='XNA']

children  days_employed  dob_years     education  education_id  \
10701         0    2358.600502         24  some college             2   

           family_status  family_status_id gender income_type  debt  \
10701  civil partnership                 1    XNA    business     0   

       total_income          purpose  
10701     32624.825  buy real estate

In [30]:
# Address the problematic values
df = df[df.gender!='XNA']

In [31]:
# Check the result - make sure it's fixed
df.gender.value_counts()

F    14164
M     7259
Name: gender, dtype: int64

In [32]:
# Let's see the values in the income_type column
df.income_type.value_counts()

employee                       11064
business                        5064
retiree                         3836
civil servant                   1453
entrepreneur                       2
unemployed                         2
paternity / maternity leave        1
student                            1
Name: income_type, dtype: int64

In [33]:
# Checking duplicates
df.duplicated().sum()

71

In [34]:
# Address the duplicates, if they exist
df = df.drop_duplicates().reset_index(drop=True)

In [35]:
# Last check whether we have any duplicates
df.duplicated().sum()

0

In [36]:
# Check the size of the dataset that you now have after your first manipulations with it
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21352 entries, 0 to 21351
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21352 non-null  int64  
 1   days_employed     19259 non-null  float64
 2   dob_years         21352 non-null  int64  
 3   education         21352 non-null  object 
 4   education_id      21352 non-null  int64  
 5   family_status     21352 non-null  object 
 6   family_status_id  21352 non-null  int64  
 7   gender            21352 non-null  object 
 8   income_type       21352 non-null  object 
 9   debt              21352 non-null  int64  
 10  total_income      19259 non-null  float64
 11  purpose           21352 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


# Working with missing values

### Restoring missing values in `total_income`

In [37]:
# Let's write a function that calculates the age category
def age_category(age):
    if age < 21:
        return 'minor'
    if 21 < age < 30:
        return 'young_adult'
    else:
        return 'adult'

In [38]:
# Test if the function works
age_category(df.loc[0]['dob_years'])

'adult'

In [39]:
# Creating new column based on function
df['age_category']=df['dob_years'].apply(age_category)

In [40]:
# Checking how values in the new column
df['age_category'].value_counts()

adult          18284
young_adult     3003
minor             65
Name: age_category, dtype: int64

In [41]:
# Replacing missing values if there are any errors
df['total_income'] = df.groupby('income_type')['total_income']\
.transform(lambda grp: grp.fillna(np.mean(grp)))

In [42]:
# Checking the number of entries in the columns
df[df['total_income'].isnull()]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose, age_category]
Index: []

In [43]:
df[df['total_income'].notna()].head()

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose age_category  
0    purchase of the house        adult  
1             car purchase        adult  
2    purchase of the house        adult  
3  supplementary education        adult  
4        to have a wedding        adult

In [44]:
df.groupby('income_type')['total_income'].mean()

income_type
business                       32397.307219
civil servant                  27361.316126
employee                       25824.679592
entrepreneur                   79866.103000
paternity / maternity leave     8612.661000
retiree                        21939.310393
student                        15712.260000
unemployed                     21014.360500
Name: total_income, dtype: float64

In [45]:
df.total_income.unique()

array([40620.102, 17932.802, 23341.752, ..., 14347.61 , 39054.888,
       13127.587])

###  Restoring values in `days_employed`

In [46]:
# Distribution of `days_employed` medians based on your identified parameters
df.groupby('income_type')['days_employed'].median()

income_type
business                         1548.009883
civil servant                    2673.404956
employee                         1576.067689
entrepreneur                      520.848083
paternity / maternity leave      3296.759962
retiree                        365176.336775
student                           578.751554
unemployed                     366413.652744
Name: days_employed, dtype: float64

In [47]:
# Distribution of `days_employed` means based on your identified parameters
df.groupby('income_type')['days_employed'].mean()

income_type
business                         2112.744402
civil servant                    3388.508552
employee                         2328.603723
entrepreneur                      520.848083
paternity / maternity leave      3296.759962
retiree                        365015.727554
student                           578.751554
unemployed                     366413.652744
Name: days_employed, dtype: float64

**We will use median.
In our case there are some extreme values that affect the mean.**

In [48]:
# Let's write a function that calculates means or medians (depending on your decision) based on your identified parameter
'check at cell below'

'check at cell below'

In [49]:
# Apply function to the income_type
df['days_employed'] = df['days_employed'].fillna(df.groupby('income_type')['days_employed'].transform('median'))

In [50]:
# Check if function worked
df[df['days_employed'].isnull()]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose, age_category]
Index: []

In [51]:
df.loc[12]

children                              0
days_employed             365176.336775
dob_years                            65
education           secondary education
education_id                          1
family_status         civil partnership
family_status_id                      1
gender                                M
income_type                     retiree
debt                                  0
total_income               21939.310393
purpose               to have a wedding
age_category                      adult
Name: 12, dtype: object

In [52]:
# Check the entries in all columns - make sure we fixed all missing values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21352 entries, 0 to 21351
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21352 non-null  int64  
 1   days_employed     21352 non-null  float64
 2   dob_years         21352 non-null  int64  
 3   education         21352 non-null  object 
 4   education_id      21352 non-null  int64  
 5   family_status     21352 non-null  object 
 6   family_status_id  21352 non-null  int64  
 7   gender            21352 non-null  object 
 8   income_type       21352 non-null  object 
 9   debt              21352 non-null  int64  
 10  total_income      21352 non-null  float64
 11  purpose           21352 non-null  object 
 12  age_category      21352 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB


In [53]:
# Check the entries in all columns - make sure we fixed all missing values
print('length of total_income column is: {}'.format(len(df['total_income'])))

length of total_income column is: 21352


In [54]:
print('length of the whole dataset is: {}'.format(len(df)))

length of the whole dataset is: 21352


Converting days_employed and total_income to int type

In [55]:
df['days_employed'].astype(int)

0          8437
1          4024
2          5623
3          4124
4        340266
          ...  
21347      4529
21348    343937
21349      2113
21350      3112
21351      1984
Name: days_employed, Length: 21352, dtype: int32

In [56]:
df['total_income'].astype(int)

0        40620
1        17932
2        23341
3        42820
4        25378
         ...  
21347    35966
21348    24959
21349    14347
21350    39054
21351    13127
Name: total_income, Length: 21352, dtype: int32

 **Conclusion of data transforming**
* We did further investigation on each column and fix multiple issues according missing values, negative value, duplicates, type converting.

**Our data is now ready to our next steps which include categorizing and further investigate for relations**

##  Categorization of data


In [57]:
# Print the values for your selected data for categorization
df.purpose

0          purchase of the house
1                   car purchase
2          purchase of the house
3        supplementary education
4              to have a wedding
                  ...           
21347       housing transactions
21348          purchase of a car
21349                   property
21350          buying my own car
21351               to buy a car
Name: purpose, Length: 21352, dtype: object

In [58]:
# Check the unique values
df.purpose.unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

**I noticed some subject that I suppose the majority of values can related to:**
1) housing<br>
2) education<br>
3) cars<br>
4) weddings<br>


In [59]:
# Let's write a function to categorize the data based on common topics
from collections import Counter

In [60]:
import nltk
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import word_tokenize

wordnet_lemma = WordNetLemmatizer()

In [61]:
lemmas_list_all = []

for purpose in df.purpose.unique():
    words = nltk.word_tokenize(purpose)
    lemmas = [wordnet_lemma.lemmatize(w, pos = 'n') for w in words]
    lemmas = [l.lower() for l in lemmas]
    for i in lemmas:
            lemmas_list_all.append(i)

In [62]:
lemmas_list_all

['purchase',
 'of',
 'the',
 'house',
 'car',
 'purchase',
 'supplementary',
 'education',
 'to',
 'have',
 'a',
 'wedding',
 'housing',
 'transaction',
 'education',
 'having',
 'a',
 'wedding',
 'purchase',
 'of',
 'the',
 'house',
 'for',
 'my',
 'family',
 'buy',
 'real',
 'estate',
 'buy',
 'commercial',
 'real',
 'estate',
 'buy',
 'residential',
 'real',
 'estate',
 'construction',
 'of',
 'own',
 'property',
 'property',
 'building',
 'a',
 'property',
 'buying',
 'a',
 'second-hand',
 'car',
 'buying',
 'my',
 'own',
 'car',
 'transaction',
 'with',
 'commercial',
 'real',
 'estate',
 'building',
 'a',
 'real',
 'estate',
 'housing',
 'transaction',
 'with',
 'my',
 'real',
 'estate',
 'car',
 'to',
 'become',
 'educated',
 'second-hand',
 'car',
 'purchase',
 'getting',
 'an',
 'education',
 'car',
 'wedding',
 'ceremony',
 'to',
 'get',
 'a',
 'supplementary',
 'education',
 'purchase',
 'of',
 'my',
 'own',
 'house',
 'real',
 'estate',
 'transaction',
 'getting',
 'higher'

In [63]:
Counter(lemmas_list_all)

Counter({'purchase': 6,
         'of': 5,
         'the': 2,
         'house': 3,
         'car': 9,
         'supplementary': 2,
         'education': 7,
         'to': 6,
         'have': 1,
         'a': 9,
         'wedding': 3,
         'housing': 3,
         'transaction': 4,
         'having': 1,
         'for': 2,
         'my': 4,
         'family': 1,
         'buy': 4,
         'real': 7,
         'estate': 7,
         'commercial': 2,
         'residential': 1,
         'construction': 1,
         'own': 4,
         'property': 4,
         'building': 2,
         'buying': 3,
         'second-hand': 2,
         'with': 2,
         'become': 1,
         'educated': 1,
         'getting': 2,
         'an': 1,
         'ceremony': 1,
         'get': 1,
         'higher': 1,
         'profile': 1,
         'university': 2,
         'renting': 1,
         'out': 1,
         'renovation': 1,
         'going': 1})

In [64]:
housing_category=['purchase','house','housing','real','estate','commercial','residential','construction','property','building','renovation']
education_category=['education','educated','university']
car_category=['car','second-hand']

In [65]:
def lemmatization_func(line):
    
    words = nltk.word_tokenize(line)
    lemmas = [wordnet_lemma.lemmatize(w, pos = 'n') for w in words]
    lemmas = [l.lower() for l in lemmas]
    
    if any(word in lemmas for word in housing_category):
        return 'housing'
    elif any(word in lemmas for word in education_category):
        return 'education'
    elif any(word in lemmas for word in car_category):
        return 'car'
    else:
        return 'other'

In [66]:
check = df.loc[0]['purpose']

In [67]:
check

'purchase of the house'

In [68]:
lemmatization_func(check)

'housing'

In [69]:
# Create a column with the categories and count the values for them
df['purpose_category'] = df['purpose'].apply(lemmatization_func)

In [70]:
df['purpose_category'].describe()

count       21352
unique          4
top       housing
freq        12160
Name: purpose_category, dtype: object

In [71]:
df['purpose_category'].value_counts()

housing      12160
education     3995
car           2887
other         2310
Name: purpose_category, dtype: int64

In [72]:
# Looking through all the numerical data in your selected column for categorization
df['total_income'].unique()

array([40620.102, 17932.802, 23341.752, ..., 14347.61 , 39054.888,
       13127.587])

In [73]:
# Getting summary statistics for the column
df['total_income'].describe()

count     21352.000000
mean      26795.979892
std       15707.677373
min        3306.762000
25%       17223.821250
50%       24291.744500
75%       32397.307219
max      362496.645000
Name: total_income, dtype: float64

**I decided to group the total income to 3 categories, based on the statistics that i got from using the describe() method (as you can see below)** 

In [74]:
# Creating function for categorizing into different numerical groups based on ranges
def income_category(income):
        if income < 17224:
            return 'bottom quarter'
        if 17224 < income < 24292:
            return 'third quarter'
        if 24292 < income < 32398:
            return 'second quarter'        
        return 'first quarter'

In [75]:
check2 = df.loc[0]['total_income']
income_category(check2)

'first quarter'

In [76]:
# Creating column with categories
df['income_category'] = df['total_income'].apply(income_category)

In [77]:
# Count each categories values to see the distribution
df['income_category'].value_counts()

second quarter    5793
third quarter     5339
bottom quarter    5338
first quarter     4882
Name: income_category, dtype: int64

## 5 Checking the Hypotheses


In [78]:
# Check the children data and paying back on time
df[df['debt']==0]['children'].value_counts()

0    12963
1     4397
2     1912
3      301
4       37
5        9
Name: children, dtype: int64

In [79]:
# Calculating default-rate based on the number of children
df[df['debt']==1]['children'].value_counts()/len(df)

0    0.049550
1    0.020701
2    0.009460
3    0.001265
4    0.000187
Name: children, dtype: float64

**Conclusion**

according to the number of children there is a small correlation between having children and paying back on time.

Relatively speaking, it seems that clients without children tend to default-rate more then clients who does have kids.


In [80]:
# Check the family status data and paying back on time
df[df['debt']==0]['family_status'].value_counts()

married              11363
civil partnership     3743
unmarried             2521
divorced              1100
widow / widower        892
Name: family_status, dtype: int64

In [81]:
# Calculating default-rate based on family status
df[df['debt']==1]['family_status'].value_counts()/len(df)

married              0.043415
civil partnership    0.018078
unmarried            0.012786
divorced             0.003981
widow / widower      0.002904
Name: family_status, dtype: float64

**Conclusion**

according to the family status there is small correlation paying back on time.

Relatively speaking, it seems that married clients tend to default-rate more then other clients.


In [82]:
# Check the income level data and paying back on time
df[df['debt']==0]['income_category'].value_counts()

second quarter    5302
bottom quarter    4913
third quarter     4866
first quarter     4538
Name: income_category, dtype: int64

In [83]:
df[df['debt']==1]['income_category'].value_counts()/len(df)

second quarter    0.022996
third quarter     0.022152
bottom quarter    0.019904
first quarter     0.016111
Name: income_category, dtype: float64

**Conclusion**

There is no pattern that points for correlation between income level and paying out on time.

In [84]:
df[df['debt']==1]['purpose_category'].value_counts()/len(df)

housing      0.043181
education    0.017329
car          0.012036
other        0.008617
Name: purpose_category, dtype: float64

**conclusion**

according to the purpose category there is small correlation paying back on time.

Relatively speaking, it seems that loans related to housing tend to default-rate more then other loans.

# General Conclusion 
At first we noticed for some technical issues that caused missing values and also an entire defective column (days_employed).

We invastigate each column seperatley in order to find problematic data and fix it (replace or delete).

After creating a nice and clean dataset we digged in further more to find correlation between different identified characteristics and paying the loans on time.

**We found out marital status have in impact on returning the loan on time:
clients how are not married tend to pay on time more then married clients.** 

**We found out that clients with children tend to pay on time more then clients without children**